unblob/handlers/archive/zip.py

```diff
@@ -98,78 +98,76 @@
+    # the file ends.
+    # E.g. in the case our firmware image looks like this:
+    # | ZIPFILE | SOMETHING ELSE | ZIPFILE |
+    # zipfile.ZipFile() will assume:
+    # |    THIS IS ALL THE SAME ZIPFILE    |
+    # For obvious reasons, this is not helpful in our case. We need to try to guess the length of
+    # the ZIP file chunk within our firmware image, independently, and then carve that chunk out.
 
     file.seek(start_offset)
-    content = io.BytesIO(file.read(zip_end - start_offset))
 
-    with ZipFile(content) as z:
-        logger.info("Found ZIP filenames", filenames=[x.filename for x in z.infolist()])
-        for g in z.infolist():
-            if g.flag_bits & 0b0001:
-                encrypted_files.add(g.filename)
-            file_names.add(g.filename)
+    # In our case, we want to find the first instance of the EOCD record header, not the last!
+    zip_end = find_first(file, b"\x50\x4b\x05\x06") + start_offset
 
-    size = zip_end - start_offset
-    return size
+    if zip_end == -1:
+        return -1
+    else:
+        file.seek(zip_end)
+        _ = cparser.end_of_central_directory(file)
+        return file.tell()
 
 
 def calculate_chunk(
     file: io.BufferedIOBase, start_offset: int
 ) -> Union[ValidChunk, UnknownChunk]:
+
     header = cparser.local_file_header(file)
     if header.version_needed_to_extract > MAXIMUM_VERSION:
         return UnknownChunk(
             start_offset=start_offset,
             reason=f"ZIP (0x{start_offset:x}): Version too high!",
         )
 
-    size = _guess_zip_size(file, start_offset)
+    end_of_zip = calculate_zipfile_end(file, start_offset)
+
+    if end_of_zip == -1:
+        return UnknownChunk(
+            start_offset=start_offset,
+            reason="Couldn't find the EOCD magic bytes.",
+        )
+
+    file.seek(start_offset)
+
+    encrypted_files = set()
+    all_files = set()
+
+    this_zip_chunk = io.BytesIO(file.read(end_of_zip - start_offset))
+    with zipfile.ZipFile(this_zip_chunk) as zip:
+        for zipinfo in zip.infolist():
+            if zipinfo.flag_bits & 0b0001:
+                encrypted_files.add(zipinfo.filename)
+            all_files.add(zipinfo.filename)
+
+    if len(encrypted_files) > 0:
+        # TODO: We can't handle encrypted ZIP files yet, so we fall back to the UnknownChunk in the
+        # cases where there are encrypted files in the ZIP.
+        return UnknownChunk(
+            start_offset=start_offset,
+            end_offset=end_of_zip,
+            reason="ZIP contains encrypted files.",
+        )
+
     return ValidChunk(
         start_offset=start_offset,
-        end_offset=start_offset + size,
+        end_offset=end_of_zip,
     )
 
 
 def make_extract_command(inpath: str, outdir: str) -> List[str]:
+    # TODO: This will just hang waiting for user input if any the ZIP is encrypted.
     return ["unzip", inpath, "-d", outdir]
```

Comment:
See the previous comment on `-P` operator to get rid of hangs when processing archives with encrypted content.

Reply:
We are using 7z with the same `-p ""`  argument.